### DISCUSSION TITLE
how to handle array of objects as combobox options

### DISCUSSION THREAD

**Discussion by stefanprobst** [ariakit/ariakit#1294] [Questions] [answered]
hi, i have an array of objects like this:

```js
const options = [{ id: 'ap', label: 'Apple' }, { id: 'or', label: 'Orange' }]
```

which i would like to use for combobox options. the combobox input and popover should display the option `label`s, however on form submit, i would like the `id` to get submitted. submitting values not in the list of options should not be allowed.

how could i do this with ariakit's combobox? (my current attempt: https://codesandbox.io/s/ariakit-combobox-wn3dn8?file=/src/App.tsx)

- example how it could be done with `headlessui` i think: https://codesandbox.io/s/headlessui-combobox-kxkc5e?file=/src/App.tsx
- `react-aria`'s combobox has a `allowsCustomValue` prop, and selects option by their `id` or any provided `key`. (although native form submit will submit the text value of the input field, not the selected id)

**Comment by diegohaz** [comment, marked as answer]
You can map the value back to its id in your submit handler.

Your CodeSandbox has an issue: you're passing a different `labels` object to `useComboboxState` on every render, which may result in unnecessary re-renders. You can either:

- Create `labels` outside of the component (which is fine since you're not using any information from the component).
- `useMemo` on `labels` (if you need any information inside the component).
- Or pass it as `defaultList` instead of `list`.

Besides that, your attempt looks fine. The hidden input is also valid if you rely on native form submission.

**Reply by stefanprobst** [reply]
thanks! do you plan on adding support for passing anything other than array of string labels as options list to combobox, or should we expect to keep handling this in userland?

**Reply by diegohaz** [reply]
There have been plans to turn `list` into an array of objects. But that made simple use cases more complex. The idea is that people who need a more complex thing will opt-out from using `list` and do the filtering themselves. `list` is there only to populate `matches`, but you can always have your own `matches` with custom logic.

We may revisit this until the final release though. Feel free to open feature requests if you have specific use cases.

**Comment by moonray** [comment]
I too need to use an value that is different from the label (e.g. `{ id: 212, label: 'abbacus' }`)
However, in my case multiple labels can have the same value; only id is unique. So converting from input text back to id seems like it wouldn't work well. Are there any other options to get the desired result?

**Reply by diegohaz** [reply]
Is your use case also in a submit handler? If so, and if there might be different items with the same value, how do you get the matching item when the user inserts a value by typing, and not by selecting it from the list?

**Reply by moonray** [reply]
In my case the user may only choose items from the list. I'm using the https://ariakit.org/examples/select-combobox (Select with Combobox) pattern.

EDIT: I do see what you're saying, and that should be a concern; likely if the item as typed does not exist, it should fail some sort of validation. My components live outside of forms, so there's no form submit handler.

EDIT: I do plan on working on another implementation using some kind of combobox pattern where a user may enter new options (my plan is to dynamically add the new item to the combobox on ENTER and sync it with an endpoint, and make that new item selected). I would run into the same discrepancy between id/label, though in some cases

**Reply by diegohaz** [reply]
SelectItem must have unique values. You can pass a custom unique value created from id and render a different children in both SelectItem and Select.

**Reply by moonray** [reply]
I have been able to make things work with SelectList and Select
However Combobox doesn't allow displaying anything different than it's value text.

**Reply by moonray** [reply]
This is what I have. It kind of works, but the selectlist isn't attached to the combobox which requires hitting tab to select an item

```ts

import {
  Combobox,
  Select,
  SelectItem,
  SelectLabel,
  SelectList,
  SelectPopover,
  useComboboxStore,
  useDisclosureStore,
  useSelectStore,
} from '@ariakit/react';
import { Observer, observer } from 'mobx-react-lite';
import React, { useContext, useState } from 'react';
import { useTranslation } from 'react-i18next';

import { Spinner } from '@feathr/components';
import { StoresContext } from '@feathr/extender/state';
import { useDebounce } from '@feathr/hooks';

import Page from '../Page';

import * as styles from './AriakitPage.css';

export interface ISelectOption {
  id: string;
  name: string;
}

function AriakitPage(): JSX.Element {
  const { Tags } = useContext(StoresContext);
  const [selected, setSelected] = useState<ISelectOption | undefined>();
  const [value, setValue] = useDebounce('');
  const { t } = useTranslation();

  const disclosure = useDisclosureStore();
  const combobox = useComboboxStore({
    disclosure,
    setValue,
  });
  const select = useSelectStore<string>({
    defaultValue: '',
    disclosure,
    setValue(newValue) {
      const tag = tags.models.find((tag) => tag.id === newValue);
      if (!tag) {
        setSelected(undefined);
      } else if (selected?.id !== newValue) {
        setSelected({ id: tag.id, name: tag.name });
      }
      const newValueForSure = tag?.name ?? newValue;
      if (value !== newValueForSure) {
        combobox.setValue(newValueForSure);
      }
    },
  });
  const selectValue = select.useState('value');

  const tags = Tags.list({
    filters: {
      name__icontains: value || undefined,
    },
    pagination: {
      page_size: 10,
      page: 0,
    },
  });

  function renderValue(id: string, name?: string): JSX.Element {
    if (name) {
      return <>{name}</>;
    }
    return <>{tags.models.find((tag) => tag.id === id)?.name ?? id}</>;
  }

  return (
    <Page
      description={t(
        'This page will  never see the light of day. It is purely an exersize in futility. But perhaps it will enlighten us?',
      )}
      title={t('Ariakit POC')}
    >
      <div className={styles.wrapper}>
        <SelectLabel className={'label'} store={select}>
          {t('Your favorite food')}
        </SelectLabel>
        <Select className={styles.select} store={select}>
          <span>
            {selected ? renderValue(selected.id, selected.name) : renderValue(selectValue)}
          </span>
        </Select>
        <SelectPopover className={styles.popover} gutter={4} sameWidth store={select}>
          <div className={styles.comboboxWrapper}>
            <Combobox
              autoSelect={true}
              className={styles.combobox}
              placeholder={t('Search...')}
              store={combobox}
            />
            {/* Prevent loss of focus by using observer to refresh loader and allow Combobox to not be refreshed */}
            <Observer>
              {function useAnonymousFunction(): JSX.Element {
                return <>{tags.isPending && <Spinner className={styles.spinner} size={20} />}</>;
              }}
            </Observer>
          </div>

          {/* Prevent loss of focus by using observer to refresh items and allow Combobox to not be refreshed */}
          <SelectList store={select}>
            <Observer>
              {function useAnonymousFunction(): JSX.Element {
                return (
                  <>
                    {tags.models.length
                      ? tags.models.map((tag) => (
                          <SelectItem className={styles.selectItem} key={tag.id} value={tag.id}>
                            {renderValue(tag.id)}
                          </SelectItem>
                        ))
                      : !tags.isPending && <div className={'no-results'}>No results found</div>}
                  </>
                );
              }}
            </Observer>
          </SelectList>
        </SelectPopover>
      </div>
    </Page>
  );
}

export default observer(AriakitPage);
```
